### PR TITLE
Resolve the slowness of HMC during init adaptation phase

### LINKED ISSUE #1673: Pyro NUTS inference slow when compared to the same PyMC3 model
For the same models below, the PyMC3 model finishes within a second, whereas the Pyro model has an extremely slow rate of progress. I came across this perf issue while working on a larger model, and isolated it to the example below. I am not sure if this is a perf bug, or whether PyMC3 uses some adaptation tricks, but it is worth looking into and learning from. The cause of the slowdown is the very small step size in Pyro (~10^-9), whereas in PyMC3, the step size throughout is of the order of ~ 10^-3. This is also not a case of bad initialization, the same can be observed for different random seeds.

Also refer to #1470. 

```python
import pymc3 as pm
import pyro

import pyro.distributions as dist
from pyro.infer.mcmc import MCMC, NUTS

import torch


def pm_model(data):
    with pm.Model() as model:
        alpha = pm.HalfCauchy('alpha', beta=2)
        beta = pm.HalfCauchy('beta', beta=2)
        l = pm.Gamma('l', alpha, beta)
        pm.Poisson('obs', l, observed=data)
    return model


def pyro_model(data):
    alpha = pyro.sample('alpha', dist.HalfCauchy(2.))
    beta = pyro.sample('beta', dist.HalfCauchy(2.))
    l = pyro.sample('l', dist.Gamma(alpha, beta))
    pyro.sample('obs', dist.Poisson(l), obs=data)


def get_samples_pymc(data, num_samples=200, warmup_steps=200):
    data = data.numpy()
    with pm_model(data):
        trace = pm.sample(draws=num_samples, tune=warmup_steps, chains=1)


def get_samples_pyro(data, num_samples=200, warmup_steps=200):
    data = data
    nuts_kernel = NUTS(pyro_model,
                       adapt_step_size=True,
                       adapt_mass_matrix=True,
                       jit_compile=True,
                       full_mass=True)
    mcmc_run = MCMC(nuts_kernel,
                    num_samples=num_samples,
                    warmup_steps=warmup_steps,
                    num_chains=1).run(data)


data = torch.tensor(4805497.)
get_samples_pymc(data)  # finishes in a second.
get_samples_pyro(data)  # orders of magnitude slower.
```

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR tries to resolve #1673 by setting a high prox_center value during the first window_adaptation. 

In addition, this also fixes some bugs:

- Missing cache for HMC. Not sure why HMC tests passed previously.
- Better definition of tree_depth (for debugging)
- Feed python float into DualAveraging scheme. Previously it is a mixed of float and tensor_scalar, which in turns forces `step_size` be a tensor_scalar, not float.

More details are in comments.

Todo:

- [x] Clean the PR (lots of printing are added for debugging)

### CODE REVIEW

**Comment by neerajprad** [comment]
> Missing fetch_from_cache for HMC. Not sure why HMC tests passed previously.

We fetch_from_cache in both, I don't see any changes related to that in this PR?

File: pyro/infer/mcmc/hmc.py [hunk 1]
```diff
@@ -307,10 +321,13 @@ def initial_trace(self):
...
         for i in range(self._max_tries_initial_trace):
             trace_log_prob_sum = self._compute_trace_log_prob(trace)
+            z = {name: node["value"].detach() for name, node in self._iter_latent_nodes(trace)}
+            #print("in configure", z)
+            #print(trace_log_prob_sum)
             if not torch_isnan(trace_log_prob_sum) and not torch_isinf(trace_log_prob_sum):
                 self._initial_trace = trace
                 return trace
-            trace = poutine.trace(self.model).get_trace(*self._args, **self._kwargs)
+            trace = poutine.trace(self.model).get_trace(self._args, self._kwargs)
```
Comment (neerajprad) [inline pyro/infer/mcmc/hmc.py]: This seems to be a regression?

File: pyro/infer/mcmc/hmc.py [hunk 2]
```diff
@@ -398,6 +415,7 @@ def sample(self, trace):
         if rand < accept_prob:
             self._accept_cnt += 1
             z = z_new
+            self._cache(potential_energy_new, z_grads_new)
```
Comment (fehiepsi) [inline pyro/infer/mcmc/hmc.py]: We should cache pe and its grad here.

File: pyro/infer/mcmc/hmc.py (hunk 1 above)
Comment (fehiepsi) [reply pyro/infer/mcmc/hmc.py]: Sorry, I'll go for a pass to clean these diffs. Currently, NUTS is fast in double precision for your example, but slow in single precision. I'm investigating it.

File: pyro/infer/mcmc/hmc.py (hunk 2 above)
Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Good catch! It is not doing any caching across sample steps since we recompute if the gradients / pe computation is None. That's why the tests weren't failing, but we weren't making use of this optimization for HMC.

**Comment by neerajprad** [comment]
This is great, @fehiepsi! Thanks for digging into this, I am taking another look now.

File: pyro/infer/mcmc/hmc.py [hunk 3]
```diff
@@ -232,11 +230,15 @@ def _find_reasonable_step_size(self, z):
...
         # keep scale step_size until accept_prob crosses its target
         # TODO: make thresholds for too small step_size or too large step_size
+        t = 0
         while direction_new == direction:
+            t += 1
             step_size = step_size_scale * step_size
-            z_new, r_new, z_grads, potential_energy = velocity_verlet(
-                z, r, self._potential_energy, self.inverse_mass_matrix, step_size)
-            energy_new = potential_energy + self._kinetic_energy(r_new)
+            r, _ = self._sample_r(name="r_presample_{}".format(t))
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Interesting, makes sense.

File: pyro/infer/mcmc/hmc.py [hunk 4]
```diff
@@ -343,38 +333,50 @@ def _initialize_model_properties(self):
...
+
+    def _initialize_sampling(self, trace):
+        z = {name: node["value"].detach() for name, node in self._iter_latent_nodes(trace)}
+        # automatically transform `z` to unconstrained space, if needed.
+        for name, transform in self.transforms.items():
+            z[name] = transform(z[name])
+        potential_energy = self._potential_energy(z)
+        self._cache(z, potential_energy, None)
+        if self._adapter.adapt_step_size:
+            self._adapter.reset_step_size_adaptation()
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Hmm..why don't we do any state setup that is trace dependent, when we set the initial trace instead? Would suggest a minor change:
 - Renaming this to `_initialize_step_size` to make the objective clearer.
 - Calling this from inside `initial_trace` instead of inside `sample`. Note that you will need to call this from both the setter and the getter.

File: pyro/infer/mcmc/nuts.py [hunk 5]
```diff
@@ -347,14 +354,18 @@ def sample(self, trace):
                         tree_weight = tree_weight + new_tree.weight
 
         if self._t < self._warmup_steps:
-            accept_prob = new_tree.sum_accept_probs / new_tree.num_proposals
+            accept_prob = sum_accept_probs / num_proposals
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply pyro/infer/mcmc/nuts.py]: Does this seem to affect any of our results?

File: pyro/infer/mcmc/hmc.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply pyro/infer/mcmc/hmc.py]: Nice solution! Thanks @neerajprad !

File: pyro/infer/mcmc/nuts.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply pyro/infer/mcmc/nuts.py]: In the old approach, if tree is diverging, accept_prob for the new_tree will be smaller than the whole tree, which will lead to a smaller step_size for the next step. However, I don't observe any improvement with this change. All tests pass just well. :(

File: tests/infer/mcmc/test_hmc.py [hunk 6]
```diff
@@ -272,7 +272,7 @@ def test_initial_trace(monkeypatch):
     true_coefs = torch.arange(1., dim + 1.)
     labels = dist.Bernoulli(logits=(true_coefs * data).sum(-1)).sample()
     # mock values to replay - from right to left
-    trace_log_prob_replay = [-5, float('NaN'), float('Inf')]
+    trace_log_prob_replay = [-5, -5, float('NaN'), float('Inf')]
```
Comment (fehiepsi) [inline tests/infer/mcmc/test_hmc.py]: we need to compute potential energy for initializing step_size.

**Comment by neerajprad** [comment]
> So in my opinion, we should encourage users reduce max_tree_depth if they want speed, and use double precision for tricky posterior.

I agree, that makes sense. We should, in that case, make `max_tree_depth` a configurable param for NUTS (in a separate PR perhaps).

File: pyro/infer/mcmc/hmc.py [hunk 7]
```diff
@@ -309,13 +299,15 @@ def initial_trace(self):
...
                 self._initial_trace = trace
+                self._initialize_step_size()  # this method also caches z and its potential energy
                 return trace
             trace = poutine.trace(self.model).get_trace(*self._args, **self._kwargs)
         raise ValueError("Model specification seems incorrect - cannot find a valid trace.")
 
     @initial_trace.setter
     def initial_trace(self, trace):
         self._initial_trace = trace
+        self._initialize_step_size()
```
Comment (neerajprad) [inline pyro/infer/mcmc/hmc.py]: Sorry, my suggestion wouldn't work because `setup` is always called after we set the `intial_trace`, but there is a much simpler solution for this. Let us remove this from `initial_trace` completely, and just put it in `setup()` towards the very end. However, you will need to make one small change to `_initialize_step_size` :

```
    def _initialize_step_size(self):
        z = {name: node["value"].detach()
             for name, node in self._iter_latent_nodes(**self.initial_trace**)}  # not self._initial_trace
        ....
```

This will set the initial trace if it is not already specified, and should be much simpler overall.

Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: The failing test should pass once you make this change.

Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Just pushed this to your branch - I hope that is okay!

Comment (fehiepsi) [reply pyro/infer/mcmc/hmc.py]: No worry, feel free to do so. :)

**Comment by fehiepsi** [comment]
I think that the last commit raises another bug when users set initial_trace after setup. The cache and step size will be set for the trace generated from `setup`, not initial_trace.

**Comment by neerajprad** [comment]
> I think that the last commit raises another bug when users set initial_trace after setup. 

This shouldn't happen though, because setup should only be called by `MCMC` and not by the user. I think we should also specify the initial trace inside as an argument to `MCMC` rather than it being a `TraceKernel` attribute, but we should make these interface changes separately.

**Comment by fehiepsi** [comment]
> we should also specify the initial trace as an argument to `MCMC` rather than it being a `TraceKernel` attribute, but we should make these interface changes separately.

Agree!
